[Bug] npm run build - does not work

### Version
1.0.0
### Reproduction link
https://creative-tim.com/product/nextjs-argon-dashboard-pro
### Operating System
Does not matter
### Device
Does not matter
### Browser & Version
Does not matter
### Steps to reproduce
1. Download a fresh copy of NextJS Argon Dashboard PRO
2. Run `npm install` and wait for the installation to complete
3. Run `npm run build` and see that it will fail with an error something like `window is not defined`
### What is expected?
It is expected that the build command would work so that the build output could be delivered on a server.
### What is actually happening?
Build command fails with an error something like `window is not defined`.

---
### Solution
1. Inside the `pages/admin/components.js` we added ad the beginning of the file, the following code:
```
if (typeof window === "undefined") {
  global.window = {};
}
if (typeof document === "undefined") {
  global.document = {};
}
```
2. Inside the `pages/admin/sortable.js` we added ad the beginning of the file, the following code:
```
if (typeof window === "undefined") {
  global.window = {};
}
if (typeof document === "undefined") {
  global.document = {};
}
```
3. Inside the `pages/admin/components.js` we replaced the following code:
```
// react plugin that creates text editor
import ReactQuill from "react-quill";
// javascript plugin that creates nice dropzones for files
import Dropzone from "dropzone";
```
With
```
import dynamic from "next/dynamic";
// react plugin that creates text editor
const ReactQuill = dynamic(() => import("react-quill"));
// javascript plugin that creates nice dropzones for files
const Dropzone = dynamic(() => import("dropzone"));
```
4. Inside the `pages/admin/sortable.js` we replaced the following code:
```
// javascript plugin that creates a sortable object from a dom object
import List from "list.js";
```
With
```
import dynamic from "next/dynamic";
// javascript plugin that creates a sortable object from a dom object
const List = dynamic(() => import("list.js"));
```
### Additional comments
Happy Hacking!
<!-- generated by creative-tim-issues. DO NOT REMOVE -->

Hey @EINazare,

Thank you for the detailed solution to the 'window not defined' issue. I ran into that issue as soon as I tried building the pro version of the Argon Dashboard. As soon as I follow the steps to solve the 'window not defined' issue, I run into another problem. I get the following error:

> nextjs-argon-dashboard-pro@1.0.0 build /Users/trex/Git/Zenda/AdminPortal/admin_portal_app
> next build

info  - Creating an optimized production build...
Warning: Built-in CSS support is being disabled due to custom CSS configuration being detected.
See here for more info: https://err.sh/next.js/built-in-css-disabled

info  - Compiled successfully
info  - Collecting page data...
info  - Generating static pages (0/28)

Error occurred prerendering page "/rtl/rtl-support". Read more: https://err.sh/next.js/prerender-error
ReferenceError: document is not defined
    at new RTL_RTL (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/server/pages/rtl/rtl-support.js:354:5)
    at d (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/react-dom/cjs/react-dom-server.node.production.min.js:36:320)
    at $a (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/react-dom/cjs/react-dom-server.node.production.min.js:39:16)
    at a.b.render (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/react-dom/cjs/react-dom-server.node.production.min.js:44:476)
    at a.b.read (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/react-dom/cjs/react-dom-server.node.production.min.js:44:18)
    at renderToString (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/react-dom/cjs/react-dom-server.node.production.min.js:54:364)
    at Object.renderPage (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/next/dist/next-server/server/render.js:54:851)
    at Function.getInitialProps (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/server/pages/_document.js:327:19)
    at loadGetInitialProps (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/next/dist/next-server/lib/utils.js:5:101)
    at renderToHTML (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/next/dist/next-server/server/render.js:54:1142)
info  - Generating static pages (7/28)
info  - Generating static pages (14/28)
info  - Generating static pages (21/28)
info  - Generating static pages (28/28)

> Build error occurred
Error: Export encountered errors on following paths:
	/rtl/rtl-support
    at exportApp (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/next/dist/export/index.js:30:1103)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async /Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/next/dist/build/index.js:39:69
    at async /Users/trex/Git/Zenda/AdminPortal/admin_portal_app/node_modules/next/dist/build/tracer.js:1:525
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! nextjs-argon-dashboard-pro@1.0.0 build: `next build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the nextjs-argon-dashboard-pro@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/trex/.npm/_logs/2021-01-23T21_46_37_359Z-debug.log


Any help will be greatly appreciated. Thank you!

Tejas

Hey Manu @EINazare,

Even with `next: 9.5.2` we're running into the issue. It is arising from additional setting in next.config.js due to our serverless deployment. When I insert `target: serverless` in next.config.js, I get the following error:
 
---------------------------------------------------------------------------------------
```
> next build

Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db
Warning: Built-in CSS support is being disabled due to custom CSS configuration being detected.
See here for more info: https://err.sh/next.js/built-in-css-disabled

info  - Creating an optimized production build  
warn  - Compiled with warnings

./node_modules/next/dist/next-server/server/load-components.js
Critical dependency: the request of a dependency is an expression

./node_modules/next/dist/next-server/server/load-components.js
Critical dependency: the request of a dependency is an expression

./node_modules/next/dist/next-server/server/require.js
Critical dependency: the request of a dependency is an expression

./node_modules/next/dist/next-server/server/require.js
Critical dependency: the request of a dependency is an expression

./node_modules/next/dist/next-server/server/require.js
Critical dependency: the request of a dependency is an expression

Warning: You have opted-out of Automatic Static Optimization due to `getInitialProps` in `pages/_app`. This does not opt-out pages with `getStaticProps`
Read more: https://err.sh/next.js/opt-out-auto-static-optimization


> Build error occurred
TypeError: Cannot read property 'style' of undefined
    at Object.p4r7 (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:23859:52)
    at __webpack_require__ (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:23:31)
    at Object.nikJ (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:23002:25)
    at __webpack_require__ (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:23:31)
    at Object.cuOD (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:17218:18)
    at __webpack_require__ (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:23:31)
    at Object.RG8o (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:10902:81)
    at __webpack_require__ (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:23:31)
    at Object.i6E2 (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:20206:87)
    at __webpack_require__ (/Users/trex/Git/Zenda/AdminPortal/admin_portal_app/.next/serverless/pages/admin/elements.js:23:31) {
  type: 'TypeError'
}
info  - Collecting page data .npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! nextjs-argon-dashboard-pro@1.0.0 build: `next build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the nextjs-argon-dashboard-pro@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/trex/.npm/_logs/2021-01-26T21_59_11_643Z-debug.log
```
---------------------------------------------------------------------------------------

On the other hand, `target: serverless` works with the free version of the Argon dashboard. Due to several advantages of the serverless architecture, it is critical that we use the build target as serverless. Can you provide any assistance with making the pro version of Argon dashboard compatible with the serverless build target? Here is my next.config.js:

```
const withPlugins = require("next-compose-plugins");
const withImages = require("next-images");
const withSass = require("@zeit/next-sass");
const withCSS = require("@zeit/next-css");
const withFonts = require("next-fonts");
const webpack = require("webpack");
const path = require("path");
// for transpiling all ESM @fullcalendar/* packages
// also, for piping fullcalendar thru babel (to learn why, see babel.config.js)
const withTM = require("next-transpile-modules")(["@fullcalendar"]);

module.exports = withFonts(
  withCSS(
    withImages(
      withSass(
        withTM({
          target: 'serverless',
          webpack(config, options) {
            config.module.rules.push({
              test: /\.(eot|ttf|woff|woff2)$/,
              use: {
                loader: "url-loader",
              },
            });
            config.resolve.modules.push(path.resolve("./"));
            return config;
          },
        })
      )
    )
  )
);

```
Best,
Tejas

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,67 @@
+## [1.1.0] 2021-02-04
+### Bug fixing
+- Changed all class components to functional ones (the `pages/_app.js` and `pages/_document.js` were kept as class components, read more here: https://nextjs.org/docs/advanced-features/custom-document)
+- https://github.com/creativetimofficial/ct-nextjs-argon-dashboard-pro/issues/1 (it also solves the serverless issues as well)
+  - please check inside `pages/admin/sortable.js` and `pages/admin/components.js` to see the solution
+### Major style changes
+- Delete the `src/assets/scss/bootstrap` folder and changed all the bootstrap imports from `src/assets/scss` files to ones from `node_modules`
+### Deleted components
+### Added components
+- `components/TagsInput/TagsInput.js` (it is actually react-tagsinput, but it uses new React API)
+### Deleted dependencies
+- @types/react
+- @types/markerclustererplus
+- @types/googlemaps
+- react-google-maps (We've replace this plugin with Google Maps API Vanilla JS)
+- react-tagsinput (we've copied the code from this plugin, and updated it to latest React version)
+### Added dependencies
++ sweetalert2@10.13.0 (so that we can import the styles from node_modules, not a downloaded version)
++ select2@4.0.13 (so that we can import the styles from node_modules, not a downloaded version)
++ @fortawesome/fontawesome-free@5.15.2 (so that we can import the styles from node_modules, not a downloaded version)
++ quill@1.3.7 (so that we can import the styles from node_modules, not a downloaded version)
++ bootstrap@4.6.0 (and deleted the downloaded version of Bootstrap)
++ node-sass-package-importer@5.3.2 (so we can import bootstrap from node_modules)
+### Updated dependencies
+```
+@fullcalendar/core            5.3.0   →    5.5.1
+@fullcalendar/daygrid         5.3.0   →    5.5.0
+@fullcalendar/interaction     5.3.0   →    5.5.0
+chart.js                      2.9.3   →    2.9.4
+list.js                       1.5.0   →    2.3.1
+moment                       2.27.0   →   2.29.1
+next                          9.5.2   →   10.0.6
+next-compose-plugins          2.2.0   →    2.2.1
+next-transpile-modules        4.1.0   →    6.1.0
+nouislider                   14.6.1   →   14.6.3
+react                       16.13.1   →   17.0.1
+react-chartjs-2              2.10.0   →   2.11.1
+react-copy-to-clipboard       5.0.2   →    5.0.3
+react-datetime               2.16.3   →    3.0.4
+react-dom                   16.13.1   →   17.0.1
+react-notification-alert     0.0.12   →   0.0.13
+react-to-print                2.9.0   →   2.12.2
+reactstrap                    8.5.1   →    8.9.0
+webpack                      4.44.1   →    4.46.0
+```
+### Warning
+_The following warnings will appear when running the installation command, but they do not affect the UI or the functionality of the product (they will be solved in our next update):_
+```
+npm WARN react-bootstrap-table2-paginator@2.1.2 requires a peer of react@^16.3.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-bootstrap-table2-paginator@2.1.2 requires a peer of react-dom@^16.3.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-bootstrap-table2-toolkit@2.1.3 requires a peer of react@^16.3.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-bootstrap-table2-toolkit@2.1.3 requires a peer of react-dom@^16.3.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-bootstrap-table-next@4.0.3 requires a peer of react@^16.3.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-bootstrap-table-next@4.0.3 requires a peer of react-dom@^16.3.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-datetime@3.0.4 requires a peer of react@^16.5.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-quill@1.3.5 requires a peer of react@^0.14.9 || ^15.3.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-select2-wrapper@1.0.4-beta6 requires a peer of react@^0.14.0 || ^15.0.0-rc || ^15.0.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-select2-wrapper@1.0.4-beta6 requires a peer of react-dom@^0.14.0 || ^15.0.0-rc || ^15.0.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
+npm WARN react-popper@1.3.7 requires a peer of react@0.14.x || ^15.0.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
+npm WARN create-react-context@0.3.0 requires a peer of react@^0.14.0 || ^15.0.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
+```
+_If they will persist in our 2.*.* version, we will drop their usages and replace them with other plugins._
+_In development mode, some of the above plugins will throw a warning because they still use React v16 syntax. If the error will persist in our 2.*.* version, we will drop their usage and replace them with other plugins._
+
 ## [1.0.0] 2020-09-14
 ### Original Release
 - Started project with NextJS

--- a/README.md
+++ b/README.md
@@ -1,7 +1,7 @@
 # [NextJS Argon Dashboard PRO](https://demos.creative-tim.com/nextjs-argon-dashboard-pro?ref=njsadp-github-readme) [![Tweet](https://img.shields.io/twitter/url/http/shields.io.svg?style=social&logo=twitter)](https://twitter.com/home?status=NextJS%20Argon%20Dashboard%20is%20a%20Premium%20NextJS%204,%20Bootstrap%204,%20React%20and%20Reactstrap%20Dashboard%20%E2%9D%A4%EF%B8%8F%0Ahttps%3A//demos.creative-tim.com/nextjs-argon-dashboard-pro-%20%23react%20%23reactstrap%20%23nextjs%20%23argon%20%23argondesign%20%23reactdashboard%20%23argonreact%20%23reactdesign%20%23bootstrap%20%23uikit%20%23freebie%20%20via%20%40CreativeTim)
 
 
-![version](https://img.shields.io/badge/version-1.0.0-blue.svg)  ![license](https://img.shields.io/badge/license-MIT-blue.svg) [![GitHub issues open](https://img.shields.io/github/issues/creativetimofficial/ct-nextjs-argon-dashboard-pro.svg?maxAge=2592000)](https://github.com/creativetimofficial/ct-nextjs-argon-dashboard-pro/issues?q=is%3Aopen+is%3Aissue) [![GitHub issues closed](https://img.shields.io/github/issues-closed-raw/creativetimofficial/ct-nextjs-argon-dashboard-pro.svg?maxAge=2592000)](https://github.com/creativetimofficial/ct-nextjs-argon-dashboard-pro/issues?q=is%3Aissue+is%3Aclosed) [![Join the chat at https://gitter.im/NIT-dgp/General](https://badges.gitter.im/NIT-dgp/General.svg)](https://gitter.im/creative-tim-general/Lobby) [![Chat](https://img.shields.io/badge/chat-on%20discord-7289da.svg)](https://discord.gg/E4aHAQy)
+![version](https://img.shields.io/badge/version-1.1.0-blue.svg)  ![license](https://img.shields.io/badge/license-MIT-blue.svg) [![GitHub issues open](https://img.shields.io/github/issues/creativetimofficial/ct-nextjs-argon-dashboard-pro.svg?maxAge=2592000)](https://github.com/creativetimofficial/ct-nextjs-argon-dashboard-pro/issues?q=is%3Aopen+is%3Aissue) [![GitHub issues closed](https://img.shields.io/github/issues-closed-raw/creativetimofficial/ct-nextjs-argon-dashboard-pro.svg?maxAge=2592000)](https://github.com/creativetimofficial/ct-nextjs-argon-dashboard-pro/issues?q=is%3Aissue+is%3Aclosed) [![Join the chat at https://gitter.im/NIT-dgp/General](https://badges.gitter.im/NIT-dgp/General.svg)](https://gitter.im/creative-tim-general/Lobby) [![Chat](https://img.shields.io/badge/chat-on%20discord-7289da.svg)](https://discord.gg/E4aHAQy)
 
 
 ![Product Image](https://raw.githubusercontent.com/creativetimofficial/public-assets/master/nextjs-argon-dashboard-pro/nextjs-argon-dashboard-pro.jpg)
@@ -313,7 +313,7 @@ We use GitHub Issues as the official bug tracker for the NextJS Argon Dashboard 
 
 ## Licensing
 
-- Copyright 2020 Creative Tim (https://www.creative-tim.com/?ref=njsadp-github-readme)
+- Copyright 2021 Creative Tim (https://www.creative-tim.com/?ref=njsadp-github-readme)
 
 ## Useful Links
 

